gateway/ln-gateway/src/lib.rs

```diff
@@ -537,105 +578,149 @@ impl Gateway {
                 }
             };
 
-            info!(
-                "Intercepting HTLC {}",
-                PrettyInterceptHtlcRequest(&htlc_request)
-            );
-            if handle.is_shutting_down() {
-                break;
-            }
+            self.handle_htlc(htlc_request, &lightning_context).await;
+        }
+    }
 
-            let payment_hash = bitcoin_hashes::sha256::Hash::from_slice(&htlc_request.payment_hash)
-                .expect("32 bytes");
-
-            // If `payment_hash` has been registered as a LNv2 payment, we try to complete
-            // the payment by getting the preimage from the federation
-            // using the LNv2 protocol. If the `payment_hash` is not registered,
-            // this HTLC is either a legacy Lightning payment or the end destination is not
-            // a Fedimint.
-            if let Ok((contract, client)) = self
-                .get_registered_incoming_contract_and_client_v2(
-                    payment_hash.to_byte_array(),
-                    htlc_request.incoming_amount_msat,
+    /// Handles an intercepted HTLC. If the HTLC is part of an incoming payment
+    /// to a federation, spawns a state machine and hands off the HTLC to it.
+    /// Otherwise, forwards the HTLC to the next hop like a normal lightning
+    /// node.
+    async fn handle_htlc(
+        &self,
+        htlc_request: InterceptHtlcRequest,
+        lightning_context: &LightningContext,
+    ) {
+        info!(
+            "Intercepting HTLC {}",
+            PrettyInterceptHtlcRequest(&htlc_request)
+        );
+
+        if self
+            .try_handle_htlc_lnv2(&htlc_request, lightning_context)
+            .await
+        {
+            return;
+        }
+
+        if self.try_handle_htlc_ln_legacy(&htlc_request).await {
+            return;
+        }
+
+        Self::forward_htlc(htlc_request, lightning_context).await;
+    }
+
+    /// Tries to handle an HTLC using the LNv2 protocol.
+    /// Returns `true` if the HTLC was handled, `false` otherwise.
+    async fn try_handle_htlc_lnv2(
+        &self,
+        htlc_request: &InterceptHtlcRequest,
+        lightning_context: &LightningContext,
+    ) -> bool {
```

Comment:
Can we use a `Result` instead of returning a boolean? I think that would be cleaner